### PR TITLE
Launch separate sbatch job to copy checkpoints over from scratch to NFS

### PR DESCRIPTION
**Pull request by suchenzang**
Decouples copying from main job, can remove gloo process group if this works.

### CODE REVIEW

**Comment by suchenzang** [comment]
TODO:
* Copy right now copies everything in local dir since there is no cleanup yet
* Fix sentinel logic to not be hardcoded in looking for 8 files per checkpoint per host
* Need to split out checkpoints by directory (splatting everything into a flat directory right now:
```
(metaseq-20221105) [susanz@<redacted> susan_batch_v13_fixed]$ ls -la
total 13706631
drwxrwxr-x  3 susanz susanz         0 Nov  6 15:52 .
drwxrwxr-x 11 susanz susanz         0 Nov  6 15:49 ..
-rw-rw-r--  1 susanz susanz       968 Nov  6 15:51 063c834d-9c0d-4b5e-bee7-b17691981649.sh
-rw-rw-r--  1 susanz susanz       968 Nov  6 15:51 97dd3148-7ed1-4e1f-8ee3-c01782bbbb2e.sh
-rw-rw-r--  1 susanz susanz 374088903 Nov  6 15:51 checkpoint_100-model_part-0-shard0.pt
-rw-rw-r--  1 susanz susanz 214312839 Nov  6 15:52 checkpoint_100-model_part-0-shard1.pt
-rw-rw-r--  1 susanz susanz 214312839 Nov  6 15:52 checkpoint_100-model_part-0-shard2.pt
-rw-rw-r--  1 susanz susanz 214312839 Nov  6 15:52 checkpoint_100-model_part-0-shard3.pt
-rw-rw-r--  1 susanz susanz 214312839 Nov  6 15:51 checkpoint_100-model_part-1-shard0.pt
-rw-rw-r--  1 susanz susanz 214312839 Nov  6 15:52 checkpoint_100-model_part-1-shard1.pt
...
```

File: metaseq/scripts/checkpoint_copy/poll_and_copy_single.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  echo "Found only $NUM_COMPLETE files so far on $(hostname)..."
+  sleep 10
+  # shellcheck disable=SC2010
+  NUM_COMPLETE=$(ls "${LOCAL_CHECKPOINT_DIR}" | grep -c "_done_checkpoint_${NUM_UPDATE}*")
+done
+
+echo -e "Found ${NUM_COMPLETE} files on $(hostname)! \n\t Proceeding to copy...  ${LOCAL_CHECKPOINT_DIR}/checkpoint_${NUM_UPDATE}* to ${NFS_CHECKPOINT_DIR}${NUM_UPDATE}/"
+
+# TODO: use rsync?
+cp "${LOCAL_CHECKPOINT_DIR}/checkpoint_${NUM_UPDATE}"* "${NFS_CHECKPOINT_DIR}${NUM_UPDATE}/"
```
Comment (Xirider) [inline metaseq/scripts/checkpoint_copy/poll_and_copy_single.sh:25]: Is checkpoint_${NUM_UPDATE} our new local checkpoint dir?

File: metaseq/scripts/checkpoint_copy/ssh_and_copy_all.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+NUM_UPDATE=$6
+
+echo $(klist)
+echo "RUNNING ssh_and_copy_all for update $NUM_UPDATE...."
+
+# Create checkpoint dir on NFS for this update
+mkdir -p "${NFS_CHECKPOINT_DIR}${NUM_UPDATE}"
+
+for HOST in $(scontrol show hostnames "$MAIN_SLURM_NODELIST") ; do
+    ssh $HOST "$OSS_DEST/metaseq/scripts/checkpoint_copy/poll_and_copy_single.sh $LOCAL_CHECKPOINT_DIR $TOTAL_FILES $NFS_CHECKPOINT_DIR $NUM_UPDATE"
```
Comment (Xirider) [inline metaseq/scripts/checkpoint_copy/ssh_and_copy_all.sh:21]: Do we need to add a time limit to `poll_and_copy_single.sh`, or will it end once `ssh_and_copy_all.sh` slurm time limit runs out?

File: metaseq/cli/train.py [hunk 3]
```diff
@@ -485,36 +486,10 @@ def post_checkpoint_callback(cfg, do_evaluate, eval_kwargs, filename):
...
-                # atomic rename of the final checkpoint directory, now that all workers have completed
-                # their copies
-                os.rename(
-                    os.path.join(destination_checkpoints_dir, temporary_checkpoint_dir),
-                    os.path.join(destination_checkpoints_dir, checkpoint_dir),
-                )
-            os.remove(filename)
+            # TODO[susanz]: Add cleanup logic.
+            file_parentpath, file_basename = os.path.split(filename)
+            done_file = os.path.join(file_parentpath, "_done_" + file_basename)
```
Comment (Xirider) [inline metaseq/cli/train.py:491]: Can we move these files it into /checkpoint_ num_update?

**Review by Xirider — APPROVED** [review]
Awesome that you figured out how to schedule all this copying!

File: metaseq/checkpoint_utils.py [hunk 4]
```diff
@@ -128,6 +139,61 @@ def _copy_if_not_async(src, dest):
...
 
 
+# nfs_dir contains trailing backslash
+SBATCH_CHECKPOINT_COPY_CMD = """#!/bin/bash
+#SBATCH --job-name=cp_{num_update}
+#SBATCH --qos=high
+#SBATCH --ntasks-per-node=1
+#SBATCH --gpus-per-node=0
+#SBATCH --nodes=1
+#SBATCH --cpus-per-task=12
```
[the remark this answers is not in the document]
Comment (suchenzang) [reply metaseq/checkpoint_utils.py:149]: These won't be on the same hosts as the main training run - separate batch job; and if not enough resources are there slurm hopefully won't schedule on the same nodes, lol.

File: metaseq/checkpoint_utils.py [hunk 5]
```diff
@@ -128,6 +139,61 @@ def _copy_if_not_async(src, dest):
...
+        if not nfs_dir.endswith("/"):
+            nfs_dir += "/"
+        copy_script_dir = os.path.join(nfs_dir, cp_script_dir)
+        if not os.path.exists(copy_script_dir):
+            os.makedirs(copy_script_dir, exist_ok=True)
+        sbatch_run_file = os.path.join(
+            copy_script_dir, f"_cp_sbatch_script_{num_update}.sh"
+        )
+        slurm_nodes = os.environ["SLURM_NODELIST"]
+        with open(sbatch_run_file, "w") as f:
```
[the remark this answers is not in the document]
Comment (suchenzang) [reply metaseq/checkpoint_utils.py:181]: There's one file per checkpoint to save - this is just tracking the script submitted each time.

File: metaseq/cli/train.py (hunk 3 above)
Comment (suchenzang) [reply metaseq/cli/train.py:491]: These are in local scratch and will be cleaned up with the checkpoints after copy to NFS is complete - no need to organize

File: metaseq/scripts/checkpoint_copy/poll_and_copy_single.sh (hunk 1 above)
Comment (suchenzang) [reply metaseq/scripts/checkpoint_copy/poll_and_copy_single.sh:25]: It's just a dir with NUM_UPDATE now 😅

File: metaseq/scripts/checkpoint_copy/ssh_and_copy_all.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+LOCAL_CHECKPOINT_DIR=$3
+TOTAL_FILES=$4
+NFS_CHECKPOINT_DIR=$5
+NUM_UPDATE=$6
+
+echo $(klist)
+echo "RUNNING ssh_and_copy_all for update $NUM_UPDATE...."
+
+# Create checkpoint dir on NFS for this update
+mkdir -p "${NFS_CHECKPOINT_DIR}${NUM_UPDATE}"
```
[the remark this answers is not in the document]
Comment (suchenzang) [reply metaseq/scripts/checkpoint_copy/ssh_and_copy_all.sh:18]: Restarts don't work properly just yet - will include this in the next PR. Good callout though!

File: metaseq/scripts/checkpoint_copy/ssh_and_copy_all.sh (hunk 2 above)
Comment (suchenzang) [reply metaseq/scripts/checkpoint_copy/ssh_and_copy_all.sh:21]: I haven't figured out the right timing thresholds yet - punting this to later, lol

File: metaseq/checkpoint_utils.py [hunk 7]
```diff
@@ -128,6 +139,61 @@ def _copy_if_not_async(src, dest):
...
+
+
+def _launch_sbatch_for_checkpoint_copy(
+    cfg: CheckpointConfig,
+    oss_dir: str,
+    num_files_per_host: int = 8,
+    num_update: int = 0,
+):
+    nfs_upload_path = (
+        cfg.cloud_upload_path if cfg.cloud_upload_path.startswith("nfs:") else None
```
[the remark this answers is not in the document]
Comment (suchenzang) [reply metaseq/checkpoint_utils.py:167]: Will be updated in our internal branch!

File: metaseq/checkpoint_utils.py [hunk 8]
```diff
@@ -128,6 +139,61 @@ def _copy_if_not_async(src, dest):
...
+#SBATCH --job-name=cp_{num_update}
+#SBATCH --qos=high
+#SBATCH --ntasks-per-node=1
+#SBATCH --gpus-per-node=0
+#SBATCH --nodes=1
+#SBATCH --cpus-per-task=12
+#SBATCH --time=4320
+#SBATCH --mem=0
+#SBATCH --output={nfs_dir}{cp_script_dir}/_cp_checkpoint_%j.stdout
+#SBATCH --error={nfs_dir}{cp_script_dir}/_cp_checkpoint_%j.stderr
```
[the remark this answers is not in the document]
Comment (suchenzang) [reply metaseq/checkpoint_utils.py:153]: Right now we have a trailing backslash that gets passed in - and line 173 makes extra sure there is a trailing slash

**Pull request merged**